### PR TITLE
Modify encoders to return torch.tensor and add pytests

### PR DESCRIPTION
**Pull request by suzannejin**

### CODE REVIEW

File: src/stimulus/data/encoding/encoders.py [hunk 1]
```diff
@@ -14,11 +15,11 @@
 class AbstractEncoder(ABC):
     """Abstract class for encoders.
 
-    Encoders are classes that encode string data into numerical representations, said numerical representations should be the exact input of the model.
+    Encoders are classes that encode the raw data into torch.tensors in meaningful ways.
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:18]: what do you mean here by "in meaningful ways"?

File: src/stimulus/data/encoding/encoders.py [hunk 2]
```diff
@@ -97,30 +100,40 @@ class TextOneHotEncoder(AbstractEncoder):
         _sequence_to_array: transforms a sequence into a numpy array
     """
 
-    def __init__(self, alphabet: str = "acgt") -> None:
+    def __init__(self, alphabet: str = "acgt", case_sensitive: bool = False, padding = True) -> None:
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:103]: Why is padding initialized to True by defaut and not False?

File: src/stimulus/data/encoding/encoders.py [hunk 3]
```diff
@@ -97,30 +100,40 @@ class TextOneHotEncoder(AbstractEncoder):
...
         self.alphabet = alphabet
+        self.case_sensitive = case_sensitive
+        self.padding = padding
+
         self.encoder = preprocessing.OneHotEncoder(
             categories=[list(alphabet)],
             handle_unknown="ignore",
         )  # handle_unknown='ignore' unsures that a vector of zeros is returned for unknown characters, such as 'Ns' in DNA sequences
 
+        # Fit the encoder with all possible characters
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:129]:
```suggestion
```

Comment (mathysgrapotte) [reply src/stimulus/data/encoding/encoders.py:129]: fit is self explanatory

File: src/stimulus/data/encoding/encoders.py [hunk 4]
```diff
@@ -130,107 +143,271 @@ def _sequence_to_array(self, sequence: str) -> np.array:
...
+                logger.error(error_msg)
+                raise ValueError(error_msg)
+            
+        # convert to torch tensor
         try:
-            return np.array(encoded_data)
-        except ValueError:
-            return encoded_data
+            return torch.from_numpy(np.array(encoded_data))
+        except Exception as e:
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:265]: in which cases would the encoding fail ? In theory, as long as the sequence are the same size, torch.from_numpy should work

File: src/stimulus/data/encoding/encoders.py [hunk 5]
```diff
@@ -130,107 +143,271 @@ def _sequence_to_array(self, sequence: str) -> np.array:
...
-        return np.array([self.encode(d) for d in data])
 
-    def decode(self, data: float) -> float:
-        """Decodes the data."""
-        return data
+        # convert to tensor
+        try:
+            return torch.tensor(data, dtype=self.dtype)
+        except (TypeError, ValueError) as e:
+            err_msg = (
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:352]: does this happen only when the elements are not floats? (two exeptions are taken into account here, TypeError and ValueError)

File: src/stimulus/data/encoding/encoders.py [hunk 6]
```diff
@@ -130,107 +143,271 @@ def _sequence_to_array(self, sequence: str) -> np.array:
...
+
+        Args:
+            data (torch.Tensor): the encoded data
+
+        Returns:
+            decoded_data (List[float]): the decoded data
+        """
+        return data.cpu().numpy().tolist()
+
+    def _check_input_dtype(self, data: Union[int, float, List[int], List[float]]) -> None:
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:370]: Here there is a _check_input_dtype function, shouldn't this be used above?

File: src/stimulus/data/encoding/encoders.py [hunk 7]
```diff
@@ -130,107 +143,271 @@ def _sequence_to_array(self, sequence: str) -> np.array:
...
+        or (not isinstance(data, list) and not isinstance(data, (int, float)))):
+            err_msg = f"Expected input data to be a float, got {type(data).__name__}"
+            logger.error(err_msg)
+            raise ValueError(err_msg)
 
 
 class IntEncoder(FloatEncoder):
     """Encoder for integer data."""
 
-    def encode(self, data: int) -> int:
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:229]: Why has the encode function for IntEncoder been removed?

**Review by mathysgrapotte — CHANGES_REQUESTED** [review]
looks great to me ! just suggested a couple of minor chages/comment changes

File: src/stimulus/data/encoding/encoders.py (hunk 6 above)
Comment (suzannejin) [reply src/stimulus/data/encoding/encoders.py:370]: since encode_all is heritated from FloatEncoder, it will be automatically used

File: src/stimulus/data/encoding/encoders.py (hunk 7 above)
Comment (suzannejin) [reply src/stimulus/data/encoding/encoders.py:229]: it gets from FloatEncoder

Comment (mathysgrapotte) [reply src/stimulus/data/encoding/encoders.py:229]: And probably encode to float32 by default ?

Comment (suzannejin) [reply src/stimulus/data/encoding/encoders.py:229]: sounds good to me

File: src/stimulus/data/encoding/encoders.py [hunk 8]
```diff
@@ -97,30 +101,38 @@ class TextOneHotEncoder(AbstractEncoder):
         _sequence_to_array: transforms a sequence into a numpy array
     """
 
-    def __init__(self, alphabet: str = "acgt") -> None:
+    def __init__(self, alphabet: str = "acgt", case_sensitive: bool = False, padding: bool = False) -> None:
```
Comment (mathysgrapotte) [inline src/stimulus/data/encoding/encoders.py:104]: replace case_sensitive by convert_lowercase to remove confusion

**Review by mathysgrapotte — CHANGES_REQUESTED** [review]
very minor change, almsot there!

**Review by mathysgrapotte — APPROVED** [review]
